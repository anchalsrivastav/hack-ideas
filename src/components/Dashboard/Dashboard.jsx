import { useLocation, useHistory } from "react-router";
import React, {useEffect, useState} from "react";
import { Tag, Row, Button, Select } from "antd";
import { HACKIDEAS, SORT_OPTIONS } from "../../constants";
import {
    addButton, cardBlock, column1, column2, content,
    dropDown,
    header,
    ideaBlock,
    row,
    title,
    upvoteButtonAndCount,
    upvoteCount
} from "./Dashboard.styles";
const {Option} = Select

const Dashboard = () => {
    const [hackList, setHackList] = useState([]);
    const history = useHistory();
    const location = useLocation();
    const userId = localStorage.getItem("userId")

    useEffect(()=>{
        let hackIdea = localStorage.getItem("hackIdea");
        if (location.state?.hackList) {
            setHackList(location.state?.hackList.length ? location.state?.hackList : [...HACKIDEAS]);
        }
        else if(hackIdea){
            hackIdea = JSON.parse(hackIdea);
            setHackList(hackIdea.length ? hackIdea : [...HACKIDEAS]);
        }
        else{
            setHackList([...HACKIDEAS])
        }
        return () => {
            localStorage.setItem("hackIdea", JSON.stringify(hackList));
        }
        // eslint-disable-next-line
    }, []);

    useEffect(() => {
        localStorage.setItem("hackIdea", JSON.stringify(hackList));
    }, [hackList]);

    const addNewHackIdea = () => {
        history.push({
            pathname: '/add',
            state: {hackList: hackList}
        })
    }

    const handleLogout = () => {
        history.push({
            pathname: '/',
            state: {hackList: hackList}
        })
    }
    
    const handleSortChange = (e) => {
        const [sortBy, sortOrder] = e.split("-")
        const hackListSortFn = (h1, h2) => {
            let a = h1[sortBy]
            let b = h2[sortBy]
            if(sortBy === "creationDate"){
                a = Date.parse(a)
                b = Date.parse(b)
            }
            return sortOrder === "asc" ? a-b : b-a
        }
        
        let hackListClone = [...hackList];
        hackListClone = hackListClone.sort(hackListSortFn)
        setHackList([...hackListClone])
    }

    const Title = ({hack}) => {
        const handleUpvote = () => {
            const newList = [...hackList]
            newList.forEach(item => {
                if(item.id === hack.id){
                    item.upvotes += 1
                }
            })
            setHackList([...newList])
        }

        return (
            <div style={title}>
                <b>{String(hack.title).toUpperCase()}</b>
                <div style={upvoteButtonAndCount}>
                    <img alt="thumbsup" src="/thumbsup.png" onClick={handleUpvote} height="100%" width="100%" />
                    <div style={upvoteCount}>({hack.upvotes})</div>
                </div>
            </div>
        )
    }

    if (userId) {
        return (
            <>
                <div style={header}>
                    <h1>HackIdeas</h1>
                    <Button onClick={handleLogout}>Logout</Button>
                </div>
                <div style={content}>
                    <Row justify="end" align="center" style={row}>
                        <h3>Sort By </h3>
                        <Select style={dropDown} onChange={handleSortChange}>
                            {Object.entries(SORT_OPTIONS).map(sortOption => <Option key={sortOption[0]} value={sortOption[0]}>{sortOption[1]}</Option>)}
                        </Select>
                    </Row>
                    <div style={addButton}><Button onClick={addNewHackIdea}> + Add HackIdea</Button></div>
                    {hackList.map((hack) => {
                        return (
                            <div style={ideaBlock} key={hack.id}>
                                <Title hack={hack}/>
                                <div style={cardBlock}>
                                    <div style={column1}><h3> Description : </h3></div>
                                    <div style={column2}><h3> {hack.description} </h3></div>
                                </div>
                                <div style={cardBlock}>
                                    <div style={column1}><h3>Tags : </h3></div>
                                    <div style={column2}>{hack.tags.map((tag) => (
                                        <Tag key={tag}>{tag}</Tag>
                                    ))}</div>
                                </div>
                            </div>
                        )
                    })}
                </div>
            </>
        )
    }
    return (<h2>User not logged In!</h2>)
}

export default Dashboard;
