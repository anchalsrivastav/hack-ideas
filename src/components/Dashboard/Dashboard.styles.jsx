export const row = {
    marginBottom: '3rem'
}

export const header = {
    backgroundColor: "#d7dbd8",
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
    paddingRight: '10px',
    paddingLeft: '10px'
}

export const content = {
    margin: 30
}

export const title = {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
    backgroundColor: "#d7dbd8",
    padding: '1rem'
}

export const dropDown = {
    width: 150,
    marginRight: '10px',
    marginLeft: '10px'
}

export const addButton = {
    display: 'flex',
    justifyContent: 'end',
    alignItems: 'center',
    marginBottom: '10px'
}

export const upvoteButtonAndCount = {
    display: 'flex',
    alignItems: 'center',
    backgroundColor: "#d7dbd8",
    marginRight: '10px',
    marginLeft: '10px'
}

export const upvoteCount = {
    marginLeft: '1rem',
}

export const ideaBlock = {
    display: 'flex',
    flexDirection: 'column',
    borderStyle: 'solid',
    borderRadius: '2px',
    marginBottom: '1rem',
    paddingBottom: '0.5rem'
}

export const cardBlock = {
    display: 'flex'
}

export const column1 = {
    flex: '1',
    marginLeft: '10%',
    paddingRight: '5%'
}

export const column2 = {
    flex: '3'
}
