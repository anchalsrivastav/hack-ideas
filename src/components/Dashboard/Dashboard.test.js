import Dashboard from "./Dashboard";
import {render} from '@testing-library/react'
import {createMemoryHistory} from 'history'
import React from 'react';
import {Router} from 'react-router-dom';
import '@testing-library/jest-dom';

localStorage.setItem("userId", "11");
const mockHistoryPush = jest.fn();

jest.mock('react-router', () => ({
    ...jest.requireActual("react-router"),
    useHistory: () => ({
        push: mockHistoryPush
    })
}));

describe('Dashboard', () => {
    beforeEach(() => {
        Object.defineProperty(window, 'matchMedia', {
            writable: true,
            value: jest.fn().mockImplementation(query => ({
                matches: false,
                media: query,
                onchange: null,
                addListener: jest.fn(),
                removeListener: jest.fn(),
                addEventListener: jest.fn(),
                removeEventListener: jest.fn(),
                dispatchEvent: jest.fn(),
            })),
        });
    });

    it("should renders dashboard component", () => {
        const history = createMemoryHistory()
        const route = '/dashboard';
        history.push(route)
        const {container, getByText} = render(
            <Router history={history}>
                <Dashboard />
            </Router>
        )
        expect(getByText('HackIdeas')).toBeInTheDocument();
        expect(getByText('TITLE 1')).toBeInTheDocument();
        expect(getByText('Description 1')).toBeInTheDocument();
        expect(getByText('feature')).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    })
})
