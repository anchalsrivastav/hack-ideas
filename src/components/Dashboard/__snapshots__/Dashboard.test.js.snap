// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard should renders dashboard component 1`] = `
<div>
  <div
    style="background-color: rgb(215, 219, 216); display: flex; justify-content: space-between; align-items: center; padding-right: 10px; padding-left: 10px;"
  >
    <h1>
      HackIdeas
    </h1>
    <button
      class="ant-btn"
      type="button"
    >
      <span>
        Logout
      </span>
    </button>
  </div>
  <div
    style="margin: 30px;"
  >
    <div
      class="ant-row ant-row-end ant-row-center"
      style="margin-bottom: 3rem;"
    >
      <h3>
        Sort By 
      </h3>
      <div
        class="ant-select ant-select-single ant-select-show-arrow"
        style="width: 150px; margin-right: 10px; margin-left: 10px;"
      >
        <div
          class="ant-select-selector"
        >
          <span
            class="ant-select-selection-search"
          >
            <input
              aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
              aria-autocomplete="list"
              aria-controls="rc_select_TEST_OR_SSR_list"
              aria-haspopup="listbox"
              aria-owns="rc_select_TEST_OR_SSR_list"
              autocomplete="off"
              class="ant-select-selection-search-input"
              id="rc_select_TEST_OR_SSR"
              readonly=""
              role="combobox"
              style="opacity: 0;"
              type="search"
              unselectable="on"
              value=""
            />
          </span>
          <span
            class="ant-select-selection-placeholder"
          />
        </div>
        <span
          aria-hidden="true"
          class="ant-select-arrow"
          style="user-select: none;"
          unselectable="on"
        >
          <span
            aria-label="down"
            class="anticon anticon-down ant-select-suffix"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="down"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
    <div
      style="display: flex; justify-content: end; align-items: center; margin-bottom: 10px;"
    >
      <button
        class="ant-btn"
        type="button"
      >
        <span>
           + Add HackIdea
        </span>
      </button>
    </div>
    <div
      style="display: flex; flex-direction: column; border-style: solid; border-radius: 2px; margin-bottom: 1rem; padding-bottom: 0.5rem;"
    >
      <div
        style="display: flex; justify-content: space-between; align-items: center; background-color: rgb(215, 219, 216); padding: 1rem;"
      >
        <b>
          TITLE 1
        </b>
        <div
          style="display: flex; align-items: center; background-color: rgb(215, 219, 216); margin-right: 10px; margin-left: 10px;"
        >
          <img
            alt="thumbsup"
            height="100%"
            src="/thumbsup.png"
            width="100%"
          />
          <div
            style="margin-left: 1rem;"
          >
            (
            0
            )
          </div>
        </div>
      </div>
      <div
        style="display: flex;"
      >
        <div
          style="flex: 1; margin-left: 10%; padding-right: 5%;"
        >
          <h3>
             Description : 
          </h3>
        </div>
        <div
          style="flex: 3;"
        >
          <h3>
             
            Description 1
             
          </h3>
        </div>
      </div>
      <div
        style="display: flex;"
      >
        <div
          style="flex: 1; margin-left: 10%; padding-right: 5%;"
        >
          <h3>
            Tags : 
          </h3>
        </div>
        <div
          style="flex: 3;"
        >
          <span
            class="ant-tag"
          >
            feature
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
